File: temperatureCalculator.cpp
#include <iostream>
#include <string>
#include <stdexcept>
#include <chrono>
#include <thread>
#include <cmath>

#define RESET   "\033[0m"
#define RED     "\033[31m"
#define YELLOW  "\033[33m"

using namespace std;

void printMenu() {
    cout << "\nOur provided operations are: "
        "\n 1. Celsius to Fahrenheit "
        "\n 2. Fahrenheit to Celsius "
        "\n 3. Celsius to Kelvin "
        "\n 4. Kelvin to Celsius "
        "\n 5. Fahrenheit to Kelvin "
        "\n 6. Kelvin to Fahrenheit "
        "\n 7. Exit\n" ;
}

float celsiusToFahrenheit(float temp){return ((temp * 9/5) + 32);}

float fahrenheitToCelsius(float temp){return ((temp - 32) * 5/9);}

int main(){
    cout << "Welcome to the Temperature Calculator";
    printMenu();

    string userOption;
    cout << "\nWhich calculation would you like to do? "; cin >> userOption;

    bool endLoop = false;

    while (!endLoop) {
        int option;

        try {
            option = stoi(userOption); // convert string to int
            if ((option < 1) || (option > 7)) {
                throw out_of_range("Number is out of range.");
            }
        }
        catch (invalid_argument&) {
            cout << RED << "\nInvalid Input. Please enter a whole number (1-7)." << RESET << endl;

            this_thread::sleep_for(chrono::seconds(2));

            printMenu();

            cout << "\nWhich calculation would you like to do? "; cin >> userOption;

            continue;
            // You can handle re-prompting the user here if needed
        }
        catch (out_of_range&) {
            cout << YELLOW << "\nNumber is out of range. Please enter a valid option (1-7)." << RESET << endl;

            this_thread::sleep_for(chrono::seconds(2));

            printMenu();

            cout << "\nWhich calculation would you like to do? "; cin >> userOption;

            continue;
        }

        if (option == 7){
            endLoop = true;
            cout << "Thank you for using Temperature Calculator";
            continue;
        }

        bool end = false;
        float temp;

        while (!end) {
            string stringTemp;
            cout << "Please input your original temperature: ";
            cin >> stringTemp;
            try {
                temp = stof(stringTemp);
                end = true;
            } catch(invalid_argument&) {
                cout << RED << "\nInvalid Temperature. Please try again.\n" << RESET;
            }
        }

        string unit;

        if (option == 1) {
            unit = "Fahrenheit";
            temp = celsiusToFahrenheit(temp);
        } else if (option == 2) {
            unit = "Celsius";
            temp = fahrenheitToCelsius(temp);
        } else if (option == 3) {
            unit = "Kelvin";
            temp += 273.15;
        } else if (option == 4) {
            unit = "Celsius";
            temp -= 273.15;
        } else if (option == 5) {
            unit = "Kelvin";
            temp = fahrenheitToCelsius(temp) + 273.15;
        } else {
            unit = "Fahrenheit";
            temp = celsiusToFahrenheit(temp - 273.15);
        }

        double result;

        if (temp == static_cast<int>(temp)) {
            // value has no decimal portion (ex: 36.0 → 36)
            result = static_cast<int>(temp);
        } else {
            // round to 4 decimal places
            result = round(temp * 10000.0) / 10000.0;
        }

        cout << "Your temperature is " << result << " degrees " << unit << "\n";

        this_thread::sleep_for(chrono::milliseconds(1500));

        printMenu();

        cout << "\nWhich calculation would you like to do? "; cin >> userOption;

    }

    return 0;
}

File: temperatureCalculator.py
import time

# ────────────────────────────────────────────
# Class to hold ANSI color codes for printing.
# Makes error/warning messages appear colored
# in the terminal for better user experience.
# ────────────────────────────────────────────
class Colors :
    RESET = '\033[0m'     # Resets text color to default
    RED = '\033[31m'      # Red text (used for errors)
    YELLOW = '\033[33m'   # Yellow text (used for warnings)


# ────────────────────────────────────────────
# Function: printMenu
# Prints the main menu of available operations.
# Keeping this as a function avoids duplication
# since we display it multiple times in the loop.
# ────────────────────────────────────────────
def printMenu() :
    print("\nOur provided operations are: "
        "\n 1. Celsius to Fahrenheit "
        "\n 2. Fahrenheit to Celsius "
        "\n 3. Celsius to Kelvin "
        "\n 4. Kelvin to Celsius "
        "\n 5. Fahrenheit to Kelvin "
        "\n 6. Kelvin to Fahrenheit "
        "\n 7. Exit")


# ────────────────────────────────────────────
# Helper conversion functions
# These functions convert temperature between
# units and return a numeric value.
# ────────────────────────────────────────────
def celsiusToFahrenheit(temp) :
    return (temp * 9/5) + 32

def fahrenheitToCelsius(temp) :
    return (temp - 32) * 5/9


print("Welcome to the Temperature Calculator")
printMenu()

# Get initial user selection (menu option)
userOption = input("\nWhich calculation would you like to do? ").strip()

# ────────────────────────────────────────────
# MAIN PROGRAM LOOP
# Continues until user chooses option 7 (Exit)
# ────────────────────────────────────────────
while True:
    # Validate user menu option
    try :
        option = int(userOption)  # Try converting to integer
    except :
        # If conversion fails, notify user and restart input
        print(Colors.RED + "\nInvalid Input. Please enter a whole number (1-7)." + Colors.RESET)
        time.sleep(2)
        printMenu()
        userOption = input("\nWhich calculation would you like to do? ").strip()
        continue

    # Check numeric range 1–7
    if option < 1 or option > 7 :
        print(Colors.YELLOW + "\nInvalid Number. Please choose a digit between 1 and 7." + Colors.RESET)
        time.sleep(2)
        printMenu()
        userOption = input("\nWhich calculation would you like to do? ").strip()
        continue

    # If user chooses Exit (option 7)
    if option == 7 :
        print("\nThank you for using this Calculator.\n")
        break

    # ────────────────────────────────────────────
    # GET TEMPERATURE INPUT
    # Allows decimals (ex: 98.6)
    # Will reprompt until valid float
    # ────────────────────────────────────────────
    while True :
        stringTemp = input("Please input your original temperature: ").strip()
        try :
            temp = float(stringTemp)  # Allows decimal temperatures
            break
        except :
            print(Colors.RED + "\nInvalid Temperature. Please try again.\n" + Colors.RESET)
            time.sleep(1)

    # ────────────────────────────────────────────
    # Perform temperature conversion based on choice
    # ────────────────────────────────────────────
    if option == 1 :
        unit = "Fahrenheit"
        temp = celsiusToFahrenheit(temp)

    elif option == 2 :
        unit = "Celsius"
        temp = fahrenheitToCelsius(temp)

    elif option == 3 :
        unit = "Kelvin"
        temp = temp + 273.15

    elif option == 4 :
        unit = "Celsius"
        temp = temp - 273.15

    elif option == 5 :
        unit = "Kelvin"
        temp = fahrenheitToCelsius(temp) + 273.15

    else :
        unit = "Fahrenheit"
        temp = celsiusToFahrenheit(temp - 273.15)

    # Format output: remove trailing '.0' for integers
    if float(temp).is_integer():
        result = int(temp)
    else:
        result = round(temp, 4)  # rounded to 4 decimals for clean output

    print(f"Your temperature is {result} degrees {unit}")
    time.sleep(1.5)

    print("\n")
    printMenu()

    # Ask user for the next conversion
    userOption = input("\nWhich calculation would you like to do next? ").strip()
